Add CardSequence for linear narrative, and allow combining with branching narratives in CardTrees

Greenheart opened the pull request:
Resolves #12 

This PR contains several related changes, split into smaller commits to hopefully simplify the reviewing process.

## How to test:
1. In `scenarios/donuts/index.ts`, import relevant data from `./tree-test.ts` and `./sequence-test.ts`.
2. For `CardTree`, the test case covers both `_startConditions` and `_endModifiers`, and seems to works as expected.
3. For `CardSequence`, the test case covers both `_startConditions` and `_endModifiers`, and seems to works as expected.

### DONE:

1. feature(cardsFromTree): support for _startConditions and _endModifiers. This will be useful for combining CardTrees within CardSequences

2. Fix bugs and improve test cases.

3. Implement CardSequence

4. Simplify CardTree (See comments in `cardsFromTree()`) 

5. Simplify CardSequence

6. Improve test case for `CardTree`. Ensure `_startConditions` and `_endModifiers` work as expected.

7. When testing is completed, remove `OLD_cardsFromTree()` implementation.

### TODO:

All done, ready for review and merge.

I think testing this with real content and improving this could be done in future PRs.

---

### How does the `CardSequence` and `CardTree` work together in practice?

Here's a drawing to explain how `CardSequence` and `CardTree` fit together

![Screenshot 2020-11-14 at 13 15 33](https://user-images.githubusercontent.com/6125097/99146803-90f8f600-267b-11eb-96a5-f1d5a6b99023.png)

On the left side of the green line, there is a `CardSequence` which consists of four `CardTree` root nodes that should appear in sequence. Everything on the right side is related to the `CardTree`s.

In `CardSequence`, we need some mechanism to control when to show the next `CardTree` root node. To accomplish this, the content utils will internally pass two extra arguments to `cardsFromTree()` when building the scenario: 1) `_startConditions` and 2) `_endModifiers`.

For example, the red dots on the right side of the green line in the drawing represent the `_endModifiers`  that will trigger the right conditions for the next `CardTree` root node of the sequence. 

Looking at the left side of the green line again, the red dot on the second `CardTree` in the `CardSequence` represents the `_startConditions` connected to the red `_endModifiers` from the previous `CardTree` in the sequence.

Similarly, the black dots work the same way to ensure the player complete the CardTree by reaching one of the `CardLeaf` nodes in the current `CardTree`, before moving on to the next one in the `CardSequence`.

---

### Notes / Ideas:

#### How to trigger the start of a `CardSequence` at a specific state?
To solve this, you could *technically* use `_startConditions`. However, I think it would be best to discourage this and only use `_startConditions` as an internal tool.

Instead, simply use `tree.conditions` within your `CardTree` root node to get the same result. This way, we keep content util internals separated from the scenario creators' own conditions and content.

#### What happens when we've merged `Card` and `EventCard` into a single type?
We will probably need to update `CardTree` and `CardSequence` to simplify and take advantage of the new card pointers.

#### Should `CardSequence` support regular `Card`s without the `CardTree` overhead?
In the future once we've merged the `Card` and `EventCard` concepts, we might want to support simpler sequences of plain `Card`s. I think there would still be an advantage to use the `CardSequence` content utils, because this would allow creators to **skip the manual references to the next cards**, and just focus on the general order by switching order within an array.

#### Then, even more interesting would be to allow combining loose and direct card pointers:

if `Card`s implement direct pointers to the next card (by specifying the cardId to show up immediately after), we could combine this with the current loose pointers in CardTree and CardSequence, implemented using flags.

Greenheart commented on scripts/content-utils/card-tree.ts:
```diff
@@ -21,7 +21,131 @@ export type CardLeaf = {
+ */
 export function cardsFromTree(
+    tree: Omit<CardTree, "modifiers">,
+    _startConditions?: WorldQuery[],
+    _endModifiers: GameWorldModifier[] = [],
```
Instead of marking these params internal by prefixing with `_`, what about if we split `cardsFromTree()` into an internal implementation and an external implementation?

By separating an internal API from the external one, maybe we could hide the internal parameters to simplify for content developers?

```ts

// External API
function cardsFromTree(tree): CardData[] {
    return INTERNAL_cardsFromTree()
}

// Internal API
function INTERNAL_cardsFromTree(tree, _startConditions, _endModifiers): CardData[] {}

```

Muthaias replied:
Exactly what I was going to suggest. Though I would prefer not to have the `INTERNAL_`-prefix. By not exporting the function it is clear that it is an internal API.

Greenheart commented on scripts/content-utils/card-tree.ts:
```diff
@@ -21,7 +21,131 @@ export type CardLeaf = {
+    // This would reduce complexity of `cardsFromTree()`, while still providing us with most of the same functionality
+    // Since these are only supposed to be used internally, I currently don't see a good use case to support multiple _startConditions or _endModifiers.
+
+    const bindRef = _startConditions?.length
+        ? _startConditions[0]["flags"] &&
+          Object.keys(_startConditions[0]["flags"])[0]
+        : undefined
```
This solution is breaking `CardTrees` nested within `CardSequences`, since the `flag` pulled from `_startConditions` will effectively be overwritten by the `CardTree` logic, which breaks the `CardSequence`.

One possible solution might be to switch back to passing down the `bindRef` as a fourth, internal argument to `cardsFromTree()`. I'll try that and see if it solves the problem

Greenheart replied:
Solved this by passing `bindRef` as the fouth internal argument. This also solved another problem that I discovered while testing the first. Big win! 🎉

Greenheart commented:
I've made great progress today! Still needs some more testing, but the current version supports the basic features we want.

Ready for early review and feedback!

Greenheart commented:
Added support for nesting `CardSequence`s as direct children of `CardTree`. This allows splitting Sequences into smaller chunks that could be reused, or combined differently to add variation.

One idea would be to generate the morning routine sequence with slight variations, where we have a sequence of two starting cards and two ending cards that are combined with random cards in the middle.

Muthaias commented on scripts/content-utils/card-sequence.ts:
```diff
@@ -0,0 +1,142 @@
+                )}\n`,
+            )
+        } else {
+            throw Error(
```
Why to you have this check? Do you expect this code to be called from javascript or any other typeless system? Otherwise the typings should be able to handle this correctly. This should never happen and if it does our types are incorrect. (Is this a case of the linter being stupid again?)

Greenheart replied:
What should we do about this? Leave it for future refactoring in another PR?

Muthaias commented on scripts/content-utils/card-sequence.ts:
```diff
@@ -0,0 +1,142 @@
+            return cardsFromSequence(node, isAvailableWhen, endModifiers)
+        } else if (isCardTree(node)) {
+            return cardsFromTree(node, isAvailableWhen, endModifiers)
+        } else {
```
Same note as previous comment on row 74

Muthaias commented on scripts/content-utils/card-tree.ts:
```diff
@@ -21,72 +21,135 @@ export type CardLeaf = {
+/**
+ * @param tree The CardTree to recursively turn into an array of cards
+ *
+ * NOTE: The following params should only be used internally by the content utils
```
I would like to avoid this `NOTE` by separating this into two different functions.

Greenheart replied:
This would be a good next PR I think.

Muthaias commented on scripts/content-utils/card-tree.ts:
```diff
@@ -21,72 +21,135 @@ export type CardLeaf = {
-    bindRef?: string,
+    _startConditions: WorldQuery[] = [],
+    _endModifiers: GameWorldModifier[] = [],
+    _bindRef?: string,
```
I was thinking you might be able to remove the need for passing `_bindRef` (which honestly is a quite obscure name) and instead pass a list of `GameWorldModifiers` that should be applied on both left and right actions. Though if it works, it works. We can refactor later. Content is priority. I am in favour of having a very relaxed approach towards content building functions.

Greenheart replied:
This could also be another PR.

Muthaias commented on scripts/scenarios/donuts/sequence-test.ts:
```diff
@@ -0,0 +1,63 @@
+    ],
+}
+
+const SEQUENCE_END_MODIFIER_RUN = propRef("SEQUENCE_END_MODIFIER_RUN")
```
Could we avoid going back to using CAPS. It hurts my eyes and makes it harder to read.

Muthaias commented on scripts/content-utils/card-sequence.ts:
```diff
@@ -0,0 +1,142 @@
+    sequence: (CardTree | CardSequence)[]
+}
+
+export function isCardSequence(node: any): node is CardSequence {
```
This type check is not implemented properly. If you take `any` as argument you need to check a lot more things. You should use `isCardSequence(node: CardSequence | CardTree)` instead.

Muthaias replied:
```
export function isCardSequence(node: CardTree | CardSequence): node is CardSequence {
    return "sequence" in node
}
```

Muthaias commented on scripts/content-utils/card-tree.ts:
```diff
@@ -21,72 +21,135 @@ export type CardLeaf = {
 }
 
+// TODO: Refactor CardTree to use this runtime type guard
+export function isCardTree(node: any): node is CardTree {
```
Pleas avoid `any`. This is not a proper typeguard unless you specify which types you expect to receive. In this case I am guessing that the type should be `CardTree | CardLeaf | CardSequence`

Muthaias replied:
```
export function isCardTree(node: CardTree | CardLeaf): node is CardTree {
    return "card" in node
}
```

Greenheart commented:
Resolved three threads with the latest updates. The remaining ones could all be solved with different PRs to allow us to merge this.

I'll resolve the conflicts and push - then we're ready to merge.

Greenheart commented:
Alright, rebased and removed the `action()` wrappers to resolve conflicts.

Builds and runs like expected.

https://github.com/Muthaias/swipeforfuture-content/pull/14/commits/0be4e0204a26aea23fe3d3c23fd65984ee8ce234#diff-03348e94e73f6972e799ee5baf2f96cc3df66f3a1023ea12f132efeaf28cf36dL70-R79

Muthaias commented:
Seems to do what it should but we should look into simplifying the implementation moving forward.
